### PR TITLE
Convert GrainTracker to C++11

### PR DESCRIPTION
**Pull request by permcody**
Convert the FeatureFloodCount and GT objects to C++11.
Also adding `std::unique` pointer helpers in DataIO.

Theres a slight reduction in code count due to this change which is nice and that's including the addition of the new helpers! I've split the commits by "conversion action taken". The only controversial part of this PR is my prolific use of `std::unique_ptr`! I'll say a few things about that. First, the use of `std::unique_ptr` in these contexts is actually perfect. There really is never sharing and in the GT ownership has, and still is intended to be transferred from the `_feature_sets` data structure to the `_unique_grains` data structure. You'll see several uses of `std::move` making that happen.

I welcome any comments from @idaholab/moose-team and @dschwen.

### CODE REVIEW

File: modules/phase_field/src/postprocessors/GrainTracker.C [hunk 1]
```diff
@@ -369,26 +357,24 @@ GrainTracker::trackGrains()
...
    * this time step. We need to figure out the rightful owner in this case and inactivate the old grain.
    */
-  for (std::map<std::pair<unsigned int, unsigned int>, std::vector<unsigned int> >::iterator it = new_grain_idx_to_existing_grain_idx.begin();
-       it != new_grain_idx_to_existing_grain_idx.end(); ++it)
-  {
-                                                                // map index     feature index
-    MooseSharedPointer<FeatureData> feature_ptr = _feature_sets[it->first.first][it->first.second];
+  for (const auto & new_to_exist_kv : new_grain_idx_to_existing_grain_idx)
+  {                                                             // map index     feature index
+    std::unique_ptr<FeatureData> feature_ptr(std::move(_feature_sets[new_to_exist_kv.first.first][new_to_exist_kv.first.second]));
```
Comment (friedmud) [inline modules/phase_field/src/postprocessors/GrainTracker.C]: I think I would prefer the `=` initialization syntax here... but it's just a preference.

Do you know of any reason to prefer one over the other?

File: modules/phase_field/src/postprocessors/GrainTracker.C [hunk 2]
```diff
@@ -485,14 +470,12 @@ GrainTracker::remapGrains()
   do
   {
     grains_remapped = false;
-    for (std::map<unsigned int, MooseSharedPointer<FeatureData> >::iterator grain_it1 = _unique_grains.begin();
-         grain_it1 != _unique_grains.end(); ++grain_it1)
+    for (auto grain_it1 = _unique_grains.begin(); grain_it1 != _unique_grains.end(); ++grain_it1)
```
Comment (friedmud) [inline modules/phase_field/src/postprocessors/GrainTracker.C]: Why not use range-based-for here?

File: modules/phase_field/src/postprocessors/GrainTracker.C [hunk 3]
```diff
@@ -485,14 +470,12 @@ GrainTracker::remapGrains()
...
     {
       if (grain_it1->second->_status == INACTIVE)
         continue;
 
-      for (std::map<unsigned int, MooseSharedPointer<FeatureData> >::iterator grain_it2 = _unique_grains.begin();
-           grain_it2 != _unique_grains.end(); ++grain_it2)
+      for (auto grain_it2 = _unique_grains.begin(); grain_it2 != _unique_grains.end(); ++grain_it2)
```
Comment (friedmud) [inline modules/phase_field/src/postprocessors/GrainTracker.C]: Same here

File: modules/phase_field/src/postprocessors/GrainTracker.C [hunk 4]
```diff
@@ -485,14 +470,12 @@ GrainTracker::remapGrains()
...
       {
         // Don't compare a grain with itself and don't try to remap inactive grains
         if (grain_it1 == grain_it2 || grain_it2->second->_status == INACTIVE)
```
Comment (friedmud) [inline modules/phase_field/src/postprocessors/GrainTracker.C]: Is this the reason (because you want to do iterator comparison?)... is there nothing that can be done with range-based-for

(BTW: I'm legitimately asking here... we're all trying to get used to the new stuff)

File: modules/phase_field/src/postprocessors/GrainTracker.C [hunk 5]
```diff
@@ -684,19 +677,19 @@ GrainTracker::attemptGrainRenumber(MooseSharedPointer<FeatureData> grain, unsign
 
     // If we reach this part of the loop, there is no simple renumbering that can be done.
     mooseAssert(_unique_grains.find(target_it->_grain_id) != _unique_grains.end(), "Error in indexing target grain in attemptGrainRenumber");
-    MooseSharedPointer<FeatureData> target_grain = _unique_grains[target_it->_grain_id];
+    FeatureData & target_grain = *_unique_grains[target_it->_grain_id];
```
Comment (friedmud) [inline modules/phase_field/src/postprocessors/GrainTracker.C]: Possibly a good point to use an `auto &` as a return value... it would help insulate this line of code from changing in the future (if you change how the data is stored again)

File: modules/phase_field/src/postprocessors/GrainTracker.C [hunk 6]
```diff
@@ -836,49 +828,46 @@ GrainTracker::updateFieldInfo()
   std::map<unsigned int, Real> tmp_map;
   MeshBase & mesh = _mesh.getMesh();
 
-  for (std::map<unsigned int, MooseSharedPointer<FeatureData> >::iterator grain_it = _unique_grains.begin(); grain_it != _unique_grains.end(); ++grain_it)
+  for (const auto & grain_kv : _unique_grains)
   {
-    unsigned int curr_var = grain_it->second->_var_idx;
+    unsigned int curr_var = grain_kv.second->_var_idx;
```
Comment (friedmud) [inline modules/phase_field/src/postprocessors/GrainTracker.C]: Another good candidate for catching by `auto`... in case the type of `_var_index` changes in the future (like to `short` or `long`)

File: modules/phase_field/src/postprocessors/GrainTracker.C (hunk 4 above)
Comment (dschwen) [inline modules/phase_field/src/postprocessors/GrainTracker.C]: You could still compare `&grain1 == &grain2` if you are using range based for with references.

😃 I love work 😃

Comment (friedmud) [inline modules/phase_field/src/postprocessors/GrainTracker.C]: @dschwen seems like a decent plan

**Comment by jwpeterson** [comment]
> NULL -> nullptr

You might want to still use libmesh_nullptr for this.  That's one of the things that the Intel compiler doesn't support very well for whatever reason...

File: modules/phase_field/src/postprocessors/GrainTracker.C (hunk 1 above)
Comment (permcody) [inline modules/phase_field/src/postprocessors/GrainTracker.C]: Nope - both are equivalent. I might switch it since this is a really busy and ugly line.

File: modules/phase_field/src/postprocessors/GrainTracker.C (hunk 4 above)
Comment (permcody) [inline modules/phase_field/src/postprocessors/GrainTracker.C]: That was indeed the reason, and yes I also thought about comparing the pointers. I'm not sure one is significantly better than the other. Note that I did lose the iterator declaration which helps a lot.

File: modules/phase_field/src/postprocessors/GrainTracker.C (hunk 5 above)
Comment (permcody) [inline modules/phase_field/src/postprocessors/GrainTracker.C]: Right, I'm still hesitantly moving to using `auto`. Scott Meyers makes a dang good case for it but it still freaks me out so I haven't put it everywhere just yet. This is probably a good candidate location.

File: modules/phase_field/src/postprocessors/GrainTracker.C (hunk 6 above)
Comment (permcody) [inline modules/phase_field/src/postprocessors/GrainTracker.C]: There are several that I've missed, mostly on purpose... Give and old C++03 programmer some time! 😆 

There is another interesting thing going on in this line, and it's more of me just being a little uncomfortable with `auto`. Note the new name of the variable. I've gone from `grain_it` to `grain_kv` (grain key/value), instead of the simpler choice of just `grain`. I did that pretty consistently in all of the map iterators. I'm not sure I love it, maybe I just need to get over it. What do you guys think?

**Comment by permcody** [comment]
@jwpeterson I wasn't aware of the `nullptr` issue. If that's the case with our minimum compiler then we need to be consistent in the framework. First are we sure it's an issue with our chosen minimum? If it is, maybe would should wrap it in a moose name and add it to our code conventions page because users will need to use this. 
What the hell Intel? More like UnIntel...

File: modules/phase_field/src/postprocessors/GrainTracker.C (hunk 6 above)
Comment (friedmud) [inline modules/phase_field/src/postprocessors/GrainTracker.C]: I've gone to doing something similar... but I typically use the word `pair`... i.e. I would call this `grain_pair`.  It is nice and literal sounding (that's exactly what it really is) and it helps me remember that I need to use `.first` and `.second` to grab values from it.

File: modules/phase_field/src/postprocessors/GrainTracker.C (hunk 4 above)
Comment (friedmud) [inline modules/phase_field/src/postprocessors/GrainTracker.C]: Yeah - neither is better than the other.  This is a good use of the iterator case.  Iterators are still very much a part of the language... and we should use them when it makes sense.

File: modules/phase_field/include/postprocessors/FauxGrainTracker.h [hunk 7]
```diff
@@ -44,14 +41,14 @@ class FauxGrainTracker : public FeatureFloodCount, public GrainTrackerInterface
    * @param show_var_coloring pass true to view variable index for a region, false for unique grain information
    * @return the entity value
    */
-  virtual Real getEntityValue(dof_id_type entity_id, FeatureFloodCount::FIELD_TYPE field_type, unsigned int var_idx) const;
+  virtual Real getEntityValue(dof_id_type entity_id, FeatureFloodCount::FIELD_TYPE field_type, unsigned int var_idx) const override;
```
Comment (dschwen) [inline modules/phase_field/include/postprocessors/FauxGrainTracker.h]: Should we be using `override` more commonly in MOOSE now?

File: modules/phase_field/include/postprocessors/FeatureFloodCount.h [hunk 8]
```diff
@@ -339,7 +339,7 @@ class FeatureFloodCount :
    * The filename and filehandle used if bubble volumes are being recorded to a file.
    * MooseSharedPointer is used so we don't have to worry about cleaning up after ourselves...
```
Comment (dschwen) [inline modules/phase_field/include/postprocessors/FeatureFloodCount.h]: update comment here

File: modules/phase_field/include/postprocessors/FeatureFloodCount.h [hunk 9]
```diff
@@ -377,8 +377,8 @@ FeatureFloodCount::writeCSVFile(const std::string file_name, const std::vector<T
 {
   if (processor_id() == 0)
   {
-    // typdef makes subsequent code easier to read...
-    typedef std::map<std::string, MooseSharedPointer<std::ofstream> >::iterator iterator_t;
+    // alias declaration
+    using iterator_t = std::map<std::string, std::unique_ptr<std::ofstream> >::iterator;
```
Comment (dschwen) [inline modules/phase_field/include/postprocessors/FeatureFloodCount.h]: why not just use auto?

File: modules/phase_field/src/postprocessors/FeatureFloodCount.C [hunk 10]
```diff
@@ -75,9 +65,9 @@ void dataLoad(std::istream & stream, FeatureFloodCount::FeatureData & feature, v
 }
 
 template<>
-void dataLoad(std::istream & stream, MooseSharedPointer<FeatureFloodCount::FeatureData> & feature, void * context)
+void dataLoad(std::istream & stream, std::unique_ptr<FeatureFloodCount::FeatureData> & feature, void * context)
```
Comment (dschwen) [inline modules/phase_field/src/postprocessors/FeatureFloodCount.C]: wait... why is this necessary? in DataIO.h you added 

```
void dataLoad(std::istream & stream, std::unique_ptr<T> & feature, void * context);
```
￼
This speacialization does nothing more than the generic template.

File: modules/phase_field/src/postprocessors/FeatureFloodCount.C [hunk 11]
```diff
@@ -738,16 +726,14 @@ FeatureFloodCount::visitElementalNeighbors(const Elem * elem, int current_idx, F
...
-       neighbor_it != all_active_neighbors.end(); ++neighbor_it)
+  for (const auto neighbor_ptr : all_active_neighbors)
   {
-    const Elem * neighbor = *neighbor_it;
-    processor_id_type my_proc_id = processor_id();
+    auto my_proc_id = processor_id();
 
     // Only recurse on elems this processor can see
-    if (neighbor && neighbor->is_semilocal(my_proc_id))
+    if (neighbor_ptr && neighbor_ptr->is_semilocal(my_proc_id))
```
Comment (dschwen) [inline modules/phase_field/src/postprocessors/FeatureFloodCount.C]: This looks familiar. Didn't I just change that to use a cached lookup? Oh, right that's not merged yet.

File: modules/phase_field/src/postprocessors/GrainTracker.C [hunk 12]
```diff
@@ -300,7 +290,7 @@ GrainTracker::trackGrains()
           {
             Moose::out << "Assigning center " << closest_match_idx << " "
                        << center_points[closest_match_idx] << " absolute distance: " << min_centroid_diff << '\n';
-            _unique_grains[closest_match_idx] = _feature_sets[map_num][feature_num];
+            _unique_grains[closest_match_idx] = std::move(_feature_sets[map_num][feature_num]);
```
Comment (dschwen) [inline modules/phase_field/src/postprocessors/GrainTracker.C]: Just to make sure I understand. After this command the state of `_feature_sets[map_num][feature_num]` is what? Invalid?

File: modules/phase_field/src/postprocessors/GrainTracker.C [hunk 13]
```diff
@@ -434,29 +420,29 @@ GrainTracker::trackGrains()
    */
   for (unsigned int map_num = 0; map_num < _maps_size; ++map_num)
     for (unsigned int i = 0; i < _feature_sets[map_num].size(); ++i)
-      if (_feature_sets[map_num][i]->_status == NOT_MARKED)
+      if (_feature_sets[map_num][i] && _feature_sets[map_num][i]->_status == NOT_MARKED)
```
Comment (dschwen) [inline modules/phase_field/src/postprocessors/GrainTracker.C]: Ok, follow up to the question above. Do you need the null check here because you may have moved ownership of teh pointer away from `_feature_sets[map_num][i]`?

File: modules/phase_field/src/postprocessors/FeatureFloodCount.C [hunk 14]
```diff
@@ -587,12 +572,11 @@ FeatureFloodCount::mergeSets(bool use_periodic_boundary_info)
...
     {
-      for (std::vector<FeatureData>::iterator it = _partial_feature_sets[rank][map_num].begin();
-           it != _partial_feature_sets[rank][map_num].end(); ++it)
+      for (auto & feature : _partial_feature_sets[rank][map_num])
       {
-        if (!it->_merged)
+        if (!feature._merged)
         {
-          _feature_sets[map_num].push_back(MooseSharedPointer<FeatureData>(new FeatureData(*it)));
+          _feature_sets[map_num].push_back(std::unique_ptr<FeatureData>(new FeatureData(feature)));
```
Comment (dschwen) [inline modules/phase_field/src/postprocessors/FeatureFloodCount.C]: Huh? Why does this even work?! Shouldn't it be `emplace_back(...)`?

Comment (dschwen) [inline modules/phase_field/src/postprocessors/FeatureFloodCount.C]: Yeah, never mind. It works because you are passing in a temporary here.

File: modules/phase_field/include/postprocessors/FauxGrainTracker.h (hunk 7 above)
Comment (permcody) [inline modules/phase_field/include/postprocessors/FauxGrainTracker.h]: Yes as we start committing new code. I don't think we necessarily need to sweep through everything but it has a lot of benefits. I think one of my favorite benefits is that it indicates that you probably don't need to comment on the overridden method.

File: modules/phase_field/include/postprocessors/FeatureFloodCount.h (hunk 9 above)
Comment (permcody) [inline modules/phase_field/include/postprocessors/FeatureFloodCount.h]: Umm, because I need it for the declaration

File: modules/phase_field/src/postprocessors/FeatureFloodCount.C (hunk 10 above)
Comment (permcody) [inline modules/phase_field/src/postprocessors/FeatureFloodCount.C]: I believe I got compile errors without it. What are you asking about? The line here, or the one in DataIO.h?

File: modules/phase_field/src/postprocessors/FeatureFloodCount.C (hunk 11 above)
Comment (permcody) [inline modules/phase_field/src/postprocessors/FeatureFloodCount.C]: Well if it would pass... 😉

File: modules/phase_field/src/postprocessors/GrainTracker.C (hunk 12 above)
Comment (permcody) [inline modules/phase_field/src/postprocessors/GrainTracker.C]: Yes, invalid, it's equivalent to having called release on it, so you can test against it like a Boolean.

File: modules/phase_field/src/postprocessors/GrainTracker.C (hunk 13 above)
Comment (permcody) [inline modules/phase_field/src/postprocessors/GrainTracker.C]: Exactly

File: modules/phase_field/include/postprocessors/FeatureFloodCount.h (hunk 9 above)
Comment (dschwen) [inline modules/phase_field/include/postprocessors/FeatureFloodCount.h]: Uhm you could use auto in that declaration below...

Comment (dschwen) [inline modules/phase_field/include/postprocessors/FeatureFloodCount.h]: Oh, the pair!

Comment (permcody) [inline modules/phase_field/include/postprocessors/FeatureFloodCount.h]: No, actually you were right. `auto` works on the pair too... Nice catch

File: modules/phase_field/src/postprocessors/FeatureFloodCount.C (hunk 10 above)
Comment (permcody) [inline modules/phase_field/src/postprocessors/FeatureFloodCount.C]: I'll look at this again - I'll see if I can remove some of these.

File: modules/phase_field/src/postprocessors/FeatureFloodCount.C [hunk 15]
```diff
@@ -867,16 +850,10 @@ FeatureFloodCount::calculateBubbleVolumes()
...
-      std::vector<MooseSharedPointer<FeatureData> >::iterator
-        bubble_it = _feature_sets[map_num].begin(),
-        bubble_end = _feature_sets[map_num].end();
-
       // Determine boundary intersection for each FeatureData object
-      for (; bubble_it != bubble_end; ++bubble_it)
-        (*bubble_it)->_intersects_boundary = setsIntersect(all_boundary_node_ids.begin(), all_boundary_node_ids.end(),
-                                                           (*bubble_it)->_local_ids.begin(), (*bubble_it)->_local_ids.end());
-    }
+      for (const auto & feature_ptr : _feature_sets[map_num])
```
Comment (permcody) [inline modules/phase_field/src/postprocessors/FeatureFloodCount.C]: Indentation is correct here. This is a nested loop, there just happens to be a comment in-between.

File: modules/phase_field/include/postprocessors/FeatureFloodCount.h (hunk 9 above)
Comment (permcody) [inline modules/phase_field/include/postprocessors/FeatureFloodCount.h]: I'm being informed that GoT is on. I guess I'd better quit this and go see if <INSERT FAVORITE CHARACTER NAME> dies tonight.

File: modules/phase_field/src/postprocessors/FeatureFloodCount.C (hunk 14 above)
Comment (friedmud) [inline modules/phase_field/src/postprocessors/FeatureFloodCount.C]: Right - it's already an rvalue... so it'll call the version of `push_back()` that takes an rvalue reference and ultimately do an rvalue assignment (calling the move operator).

That said - I _do_ think we should get used to using `emplace_back()`.  It can be a good optimization.  I actually have a `push_back`/`emplace_back` in a critical section in my new code where I'm creating shared pointers and putting them on the end of a `std::vector` and `emplace_back()` gave a measurable speed improvement (in my case I have a good upper bound for the number of things in that vector so I'm able to do a good `reserve()` up front to eliminate any speed loss due to re-allocation).

File: modules/phase_field/include/postprocessors/FeatureFloodCount.h (hunk 9 above)
Comment (friedmud) [inline modules/phase_field/include/postprocessors/FeatureFloodCount.h]: Hah - it really is too bad that they had to kill Jon Snow again...

Comment (dschwen) [inline modules/phase_field/include/postprocessors/FeatureFloodCount.h]: Uuuuuaaaaarrrrggggghhhhhhhhhb

Comment (mangerij) [inline modules/phase_field/include/postprocessors/FeatureFloodCount.h]: Hodor.

File: modules/phase_field/src/postprocessors/FeatureFloodCount.C [hunk 16]
```diff
@@ -702,12 +688,12 @@ FeatureFloodCount::flood(const DofObject * dof_object, int current_idx, FeatureD
    * If we reach this point (i.e. we haven't returned early from this routine),
    * we've found a new mesh entity that's part of a feature.
    */
-  unsigned int map_num = _single_map_mode ? 0 : current_idx;
-  processor_id_type rank = processor_id();
+  auto map_num = _single_map_mode ? decltype(current_idx)(0) : current_idx;
```
Comment (friedmud) [inline modules/phase_field/src/postprocessors/FeatureFloodCount.C]: OOooooooh - nice use of `decletype()`!!

File: modules/phase_field/src/postprocessors/FeatureFloodCount.C [hunk 17]
```diff
@@ -702,12 +688,12 @@ FeatureFloodCount::flood(const DofObject * dof_object, int current_idx, FeatureD
...
+  auto rank = processor_id();
 
   // New Feature (we need to create it and add it to our data structure)
   if (!feature)
-    _partial_feature_sets[rank][map_num].push_back(FeatureData(current_idx));
+    _partial_feature_sets[rank][map_num].emplace_back(FeatureData(current_idx));
```
Comment (friedmud) [inline modules/phase_field/src/postprocessors/FeatureFloodCount.C]: This isn't really the right way to use `emplace_back()`.  It will compile and work... but I won't give you the "perfect forwarding in-place construction".  What it's actually doing is creating a `FeatureData` and then (since it's an rvalue) it will be rvalue assigned (moved) into the memory spot (or, if there is no move operator for this object then it will be _copied_ into place... just like you would get with `push_back()`!!)

With `emplace_back()` you use "variadic templates" to actually just pass through the constructor arguments.  So this line of code should really look like:

```C++
_partial_feature_sets[rank][map_num].emplace_back(current_idx);
```

That `current_idx` will get passed through to the constructor and the object will actually just get created _once_ right in the right memory spot.

A _really_ good resource for reading about `emplace_back()` is here: http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2007/n2345.pdf

(seriously - it's a good read... not just a technical document).

File: modules/phase_field/src/postprocessors/FeatureFloodCount.C [hunk 18]
```diff
@@ -579,20 +562,19 @@ FeatureFloodCount::mergeSets(bool use_periodic_boundary_info)
...
     {
-      for (std::vector<FeatureData>::iterator it = _partial_feature_sets[rank][map_num].begin();
-           it != _partial_feature_sets[rank][map_num].end(); ++it)
+      for (auto & feature : _partial_feature_sets[rank][map_num])
       {
-        if (!it->_merged)
+        if (!feature._merged)
         {
-          _feature_sets[map_num].push_back(MooseSharedPointer<FeatureData>(new FeatureData(*it)));
+          _feature_sets[map_num].emplace_back(std::unique_ptr<FeatureData>(new FeatureData(feature)));
```
Comment (friedmud) [inline modules/phase_field/src/postprocessors/FeatureFloodCount.C]: BTW: Despite what I said below about how to use `emplace_back()` this _is_ the best you can do for putting pointers into a `std::vector` using C++11.  Even in C++14 you still want to wrap the pointer in `make_unique()` like so:

```C++
_feature_sets[map_num].emplace_back(std:make_unique<FeatureData>(feature));
```

Either way you _don't_ want to try passing a raw pointer in to be forwarded like this:

```C++
_feature_sets[map_num].emplace_back(new FeatureData(feature));
```

The reason is that the pointer will be "discarded" if the vector throws an exception... which means it will leak memory.

File: modules/phase_field/src/postprocessors/FeatureFloodCount.C (hunk 17 above)
Comment (friedmud) [inline modules/phase_field/src/postprocessors/FeatureFloodCount.C]: Clarification... I said:

> it will be rvalue assigned (moved) into the memory spot (or, if there is no move operator for this object then it will be copied into place

I just wanted to be clear that `emplace_back()` is passing the arguments through to the constructors of `FeatureData`.  So when I say "rvalue assigned" I mean that it will attempt to call a constructor that takes a `FeatureData &&` (called a "move constructor") and if one doesn't exist then it will try to call a constructor that takes `FeatureData &` (ie... copy constructor).

In this case it looks like `FeatureData` doesn't have move constructor so this line of code as it currently stands will be effectively doing the same thing `push_back()` would do...

Comment (permcody) [inline modules/phase_field/src/postprocessors/FeatureFloodCount.C]: Oh I see. Thanks! I'll give that paper a read

File: modules/phase_field/src/postprocessors/FeatureFloodCount.C (hunk 18 above)
Comment (permcody) [inline modules/phase_field/src/postprocessors/FeatureFloodCount.C]: Right, I read about that in Meyer's book. I commented on the libMesh issue to have John or Roy add make_unique wrapper to libMesh so we don't have to wait several years to start using that feature.

**Comment by friedmud** [comment]
BTW: You should make sure that all of your objects have move constructors.

In particular, `FeatureData` has a TON of large objects inside of it and you are putting it directly in a `std::vector`!  Since you only have a copy constructor for it (and it is explicitly defined) there is currently no move constructor so each time the vector needs to reallocate because it's growing it has to copy all od that data!

You should either define a move constructor... or, if you think it will do the right thing, just declare that a default move constructor can be used like so:

```C++
FeatureData(FeatureData &&) = default;
```

This could be a pretty large speedup to your code here!

**Comment by permcody** [comment]
I'll look into that too. I'll be interested in performing some timings after each of these optimizations to see if it makes much of a difference. Even if it doesn't, there's no reason not to have these new members.

**Comment by dschwen** [comment]
Is this good to go now, @permcody ?

**Comment by permcody** [comment]
Yeah - I'm happy with it. I'm sure I'll continue to mod it going forward but you can merge this.

**Comment by dschwen** [comment]
:+1: I'll merge this and rebase my PR onto it
